File: framework/src/main/java/com/linxiao/framework/widget/ObservableScrollView.java
package com.linxiao.framework.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.ScrollView;

/**
 * 设置滚动距离监听的ScrollView
 * <p>通过设置OnScrollChangedListener监听滚动距离</p>
 *
 * @author linxiao
 * @version 1.0
 */
public class ObservableScrollView extends ScrollView {

    public interface OnScrollChangedListener {

        void onScrollChanged(ScrollView scrollView, int scrolledX, int scrolledY, int dx, int dy);

    }

    private OnScrollChangedListener listener;

    public ObservableScrollView(Context context) {
        super(context);
    }

    public ObservableScrollView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public ObservableScrollView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public void setOnScrollChangedListener(OnScrollChangedListener listener) {
        this.listener = listener;
    }

    @Override
    protected void onScrollChanged(int l, int t, int oldl, int oldt) {
        super.onScrollChanged(l, t, oldl, oldt);
        int dl = l - oldl;
        int dt = t - oldt;;
        if(listener != null) {
            listener.onScrollChanged(this, getScrollX(), getScrollY(), dl, dt);
        }

    }
}

File: framework/src/main/java/com/linxiao/framework/widget/ObservableScrollView.kt
package com.linxiao.framework.widget

import android.content.Context
import android.util.AttributeSet
import android.widget.RelativeLayout
import android.widget.ScrollView
import androidx.core.widget.NestedScrollView

/**
 * 设置滚动距离监听的ScrollView
 *
 * 通过设置OnScrollChangedListener监听滚动距离
 *
 * @author linxiao
 * @version 1.0
 */
class ObservableScrollView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : NestedScrollView(context, attrs, defStyleAttr) {
    interface OnScrollChangedListener {
        fun onScrollChanged(
            scrollView: NestedScrollView?,
            scrolledX: Int,
            scrolledY: Int,
            dx: Int,
            dy: Int
        )
    }

    private var listener: OnScrollChangedListener? = null

    fun setOnScrollChangedListener(listener: OnScrollChangedListener?) {
        this.listener = listener
    }

    override fun onScrollChanged(l: Int, t: Int, oldl: Int, oldt: Int) {
        super.onScrollChanged(l, t, oldl, oldt)
        val dl = l - oldl
        val dt = t - oldt
        listener?.onScrollChanged(this, scrollX, scrollY, dl, dt)
    }
}
